logdna/logdna.py

```diff
@@ -39,88 +43,93 @@ def __init__(self, key, options={}):
 
         if isinstance(self.tags, str):
             self.tags = [tag.strip() for tag in self.tags.split(',')]
         elif not isinstance(self.tags, list):
             self.tags = []
-
-
         self.setLevel(logging.DEBUG)
         self.lock = threading.RLock()
 
-        self.session = requests.Session()
-        retry = Retry(connect=self.connection_retries, backoff_factor=self.retry_backoff_factor)
-        adapter = HTTPAdapter(max_retries=retry)
-        self.session.mount('https://', adapter)
-
     def buffer_log(self, message):
         if message and message['line']:
             if len(message['line']) > self.max_length:
                 message['line'] = message['line'][:self.max_length] + ' (cut off, too long...)'
                 if self.verbose in ['true', 'debug', 'd']:
-                    logger.debug('Line was longer than %s chars and was truncated.', str(self.max_length))
-
-        self.buf_byte_length += sys.getsizeof(message)
+                    self.internalLogger.debug('Line was longer than %s chars and was truncated.', self.max_length)
 
         # Attempt to acquire lock to write to buf, otherwise write to secondary as flush occurs
-        if not self.lock.acquire(blocking=False):
-            self.secondary.append(message)
-        else:
-            self.buf.append(message)
+        if self.lock.acquire(blocking=False):
+            lengths_of_messages = list(map(lambda mes: len(mes['line']), self.buf))
+            buf_size = reduce(lambda x, y: x + len(y['line']), self.buf, 0)
+
+            if buf_size + len(message['line']) < self.buf_retention_limit:
+                self.buf.append(message)
+            else:
+                self.internalLogger.debug('The buffer size exceeded the limit: %s', self.buf_retention_limit)
             self.lock.release()
-            if self.buf_byte_length >= self.flush_limit:
+
+            if buf_size >= self.flush_limit and not self.exception_flag:
                 self.flush()
-                return
+        else:
+            self.secondary.append(message)
 
         if not self.flusher:
-            self.flusher = threading.Timer(self.flush_interval, self.flush)
+            interval = self.retry_interval_secs if self.exception_flag else self.flush_interval
+            self.flusher = threading.Timer(interval, self.flush)
             self.flusher.start()
 
+    def cleaning_after_success(self):
+        del self.buf[:]
+        self.exception_flag = False
+        if self.flusher:
+            self.flusher.cancel()
+            self.flusher = None
+
+    def excpetion_actions(self, excpetion):
+       if self.flusher:
+           self.flusher.cancel()
+           self.flusher = None
+       self.exception_flag = True
+       if self.verbose in ['true', 'error', 'err', 'e']:
+           self.internalLogger.debug('Error sending logs %s', excpetion)
+
+    # do not call without aquiring the lock
+    def send_request(self):
+       self.buf = self.buf + self.secondary
```

Comment:
Since you're adding to an existing list, `self.buf.extend(self.secondary)` is probably better.